### ISSUE TITLE
incorrect diagnostic location on string templates for the same variable

### ISSUE THREAD

**Issue by OmarTawfik** [rust-lang/rust#109928]
I tried this code:

```rust
fn main() {
    let c = 'c';
    println!("{c} {c:X}");
}
```

I expected to see this happen: the error below would be shown on `{c:X}`:

```log
3 |     println!("{c} {c:X}");
  |                    ^ the trait `UpperHex` is not implemented for `char`
```

Instead, this happened: the error was shown on `{c}`:

```log
3 |     println!("{c} {c:X}");
  |                ^ the trait `UpperHex` is not implemented for `char`
```

Repro in playground:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn+main%28%29+%7B%0A++++let+c+%3D+%27c%27%3B%0A++++println%21%28%22%7Bc%7D+%7Bc%3AX%7D%22%29%3B%0A%7D

**Comment by jyn514** [comment]
Current output on nightly:
```
error[E0277]: the trait bound `char: UpperHex` is not satisfied
 --> src/main.rs:3:19
  |
3 |     println!("{c} {c:X}");
  |                   ^^^^^ the trait `UpperHex` is not implemented for `char`
```
